File: app/src/main/java/com/hippo/ehviewer/ui/scene/LRRCategoriesScene.java
package com.hippo.ehviewer.ui.scene;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.RadioGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.appbar.MaterialToolbar;
import com.google.android.material.switchmaterial.SwitchMaterial;
import com.hippo.ehviewer.EhApplication;
import com.hippo.ehviewer.ServiceRegistry;
import com.hippo.ehviewer.R;
import com.hippo.ehviewer.client.data.ListUrlBuilder;
import com.hippo.ehviewer.client.lrr.LRRAuthManager;
import com.hippo.ehviewer.client.lrr.LRRApiUtilsKt;
import com.hippo.ehviewer.client.lrr.LRRCategoryApi;
import com.hippo.ehviewer.client.lrr.data.LRRCategory;
import com.hippo.ehviewer.ui.scene.gallery.list.GalleryListScene;
import com.hippo.scene.Announcer;
import com.hippo.ehviewer.client.lrr.LRRCoroutineHelper;
import com.hippo.util.IoThreadPoolExecutor;

import java.util.ArrayList;
import java.util.List;

import okhttp3.OkHttpClient;

/**
 * Scene that displays LANraragi categories with full CRUD support.
 * Pinned categories appear first. Clicking a category navigates
 * to GalleryListScene filtered by that category.
 */
public class LRRCategoriesScene extends BaseScene {

    private static final String TAG = "LRRCategoriesScene";

    private RecyclerView mRecyclerView;
    private View mProgress;
    private View mErrorView;
    private ImageView mErrorIcon;
    private TextView mErrorTitle;
    private TextView mErrorMessage;
    private Button mErrorRetry;
    private MaterialToolbar mToolbar;

    private List<LRRCategory> mCategories = new ArrayList<>();
    private CategoryAdapter mAdapter;

    @Override
    public int getNavCheckedItem() {
        return R.id.nav_favourite;
    }

    @Nullable
    @Override
    public View onCreateView2(@NonNull LayoutInflater inflater,
                              @Nullable ViewGroup container,
                              @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.scene_lrr_categories, container, false);

        mToolbar = view.findViewById(R.id.toolbar);
        mProgress = view.findViewById(R.id.progress);
        mErrorView = view.findViewById(R.id.error_view);
        mErrorIcon = view.findViewById(R.id.error_icon);
        mErrorTitle = view.findViewById(R.id.error_title);
        mErrorMessage = view.findViewById(R.id.error_message);
        mErrorRetry = view.findViewById(R.id.error_retry);
        mRecyclerView = view.findViewById(R.id.recycler_view);

        mToolbar.setTitle(R.string.lrr_categories_title);
        mToolbar.setNavigationIcon(R.drawable.v_arrow_left_dark_x24);
        mToolbar.setNavigationOnClickListener(v -> onBackPressed());

        // Add "+" menu item
        mToolbar.inflateMenu(R.menu.scene_lrr_categories);
        mToolbar.setOnMenuItemClickListener(item -> {
            if (item.getItemId() == R.id.action_add_category) {
                showCategoryDialog(null);
                return true;
            }
            return false;
        });

        mAdapter = new CategoryAdapter();
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        mRecyclerView.setHasFixedSize(true);
        mRecyclerView.setAdapter(mAdapter);

        fetchCategories();

        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mRecyclerView = null;
        mProgress = null;
        mErrorView = null;
        mErrorIcon = null;
        mErrorTitle = null;
        mErrorMessage = null;
        mErrorRetry = null;
        mToolbar = null;
        mAdapter = null;
    }

    // ==================== Data Loading ====================

    private void fetchCategories() {
        showProgress();

        IoThreadPoolExecutor.Companion.getInstance().execute(() -> {
            try {
                Context ctx = getEHContext();
                if (ctx == null) return;
                String serverUrl = LRRAuthManager.getServerUrl();
                OkHttpClient client = ServiceRegistry.INSTANCE.getNetworkModule().getOkHttpClient();
                @SuppressWarnings("unchecked")
                List<LRRCategory> categories = (List<LRRCategory>) LRRCoroutineHelper.runSuspend(
                        (scope, cont) -> LRRCategoryApi.getCategories(client, serverUrl, cont)
                );

                // Sort: pinned first, then by name
                List<LRRCategory> pinned = new ArrayList<>();
                List<LRRCategory> unpinned = new ArrayList<>();
                for (LRRCategory cat : categories) {
                    if (cat.name == null || cat.name.isEmpty()) continue;
                    if (cat.isPinned()) {
                        pinned.add(cat);
                    } else {
                        unpinned.add(cat);
                    }
                }
                pinned.addAll(unpinned);

                if (getActivity() == null) return;
                getActivity().runOnUiThread(() -> {
                    mCategories.clear();
                    mCategories.addAll(pinned);
                    if (mCategories.isEmpty()) {
                        showEmpty(getString(R.string.lrr_categories_empty));
                    } else {
                        showList();
                        if (mAdapter != null) mAdapter.notifyDataSetChanged();
                    }
                });
            } catch (Exception e) {
                Log.e(TAG, "Failed to load categories", e);
                if (getActivity() == null) return;
                getActivity().runOnUiThread(() ->
                        showError(LRRApiUtilsKt.friendlyError(e)));
            }
        });
    }

    // ==================== CRUD Operations ====================

    /**
     * Show create/edit category dialog.
     * @param category null for create, non-null for edit.
     */
    private void showCategoryDialog(@Nullable LRRCategory category) {
        if (getEHContext() == null) return;
        boolean isEdit = category != null;

        View dialogView = LayoutInflater.from(getEHContext()).inflate(R.layout.dialog_category_edit, null);
        EditText nameInput = dialogView.findViewById(R.id.category_name);
        RadioGroup typeGroup = dialogView.findViewById(R.id.category_type);
        View searchLayout = dialogView.findViewById(R.id.search_layout);
        EditText searchInput = dialogView.findViewById(R.id.category_search);
        SwitchMaterial pinnedSwitch = dialogView.findViewById(R.id.category_pinned);

        // Pre-fill for edit mode
        if (isEdit) {
            nameInput.setText(category.name);
            pinnedSwitch.setChecked(category.isPinned());
            if (category.isDynamic()) {
                typeGroup.check(R.id.type_dynamic);
                searchLayout.setVisibility(View.VISIBLE);
                searchInput.setText(category.search);
            }
            // Disable type switch for edit (can't change static↔dynamic if archives exist)
            if (category.archives != null && !category.archives.isEmpty()) {
                dialogView.findViewById(R.id.type_static).setEnabled(false);
                dialogView.findViewById(R.id.type_dynamic).setEnabled(false);
            }
        }

        // Toggle search field visibility
        typeGroup.setOnCheckedChangeListener((group, id) ->
                searchLayout.setVisibility(id == R.id.type_dynamic ? View.VISIBLE : View.GONE));

        new AlertDialog.Builder(getEHContext())
                .setTitle(isEdit ? R.string.lrr_category_edit : R.string.lrr_category_create)
                .setView(dialogView)
                .setPositiveButton(android.R.string.ok, (dialog, which) -> {
                    String name = nameInput.getText().toString().trim();
                    if (name.isEmpty()) {
                        Toast.makeText(getEHContext(), R.string.lrr_category_name_empty, Toast.LENGTH_SHORT).show();
                        return;
                    }
                    boolean isDynamic = typeGroup.getCheckedRadioButtonId() == R.id.type_dynamic;
                    String search = isDynamic ? searchInput.getText().toString().trim() : null;
                    boolean pinned = pinnedSwitch.isChecked();

                    if (isEdit) {
                        updateCategory(category.id, name, search, pinned);
                    } else {
                        createCategory(name, search, pinned);
                    }
                })
                .setNegativeButton(android.R.string.cancel, null)
                .show();
    }

    private void createCategory(String name, @Nullable String search, boolean pinned) {
        IoThreadPoolExecutor.Companion.getInstance().execute(() -> {
            try {
                Context ctx = getEHContext();
                if (ctx == null) return;
                String serverUrl = LRRAuthManager.getServerUrl();
                OkHttpClient client = ServiceRegistry.INSTANCE.getNetworkModule().getOkHttpClient();
                LRRCoroutineHelper.runSuspend(
                        (scope, cont) -> LRRCategoryApi.createCategory(client, serverUrl, name, search, pinned, cont));
                if (getActivity() == null) return;
                getActivity().runOnUiThread(() -> {
                    Toast.makeText(getEHContext(), R.string.lrr_category_created, Toast.LENGTH_SHORT).show();
                    fetchCategories();
                });
            } catch (Exception e) {
                Log.e(TAG, "Failed to create category", e);
                if (getActivity() == null) return;
                getActivity().runOnUiThread(() ->
                        Toast.makeText(getEHContext(), LRRApiUtilsKt.friendlyError(e), Toast.LENGTH_SHORT).show());
            }
        });
    }

    private void updateCategory(String categoryId, String name, @Nullable String search, boolean pinned) {
        IoThreadPoolExecutor.Companion.getInstance().execute(() -> {
            try {
                Context ctx = getEHContext();
                if (ctx == null) return;
                String serverUrl = LRRAuthManager.getServerUrl();
                OkHttpClient client = ServiceRegistry.INSTANCE.getNetworkModule().getOkHttpClient();
                LRRCoroutineHelper.runSuspend(
                        (scope, cont) -> LRRCategoryApi.updateCategory(client, serverUrl, categoryId, name, search, pinned, cont));
                if (getActivity() == null) return;
                getActivity().runOnUiThread(() -> {
                    Toast.makeText(getEHContext(), R.string.lrr_category_updated, Toast.LENGTH_SHORT).show();
                    fetchCategories();
                });
            } catch (Exception e) {
                Log.e(TAG, "Failed to update category", e);
                if (getActivity() == null) return;
                getActivity().runOnUiThread(() ->
                        Toast.makeText(getEHContext(), LRRApiUtilsKt.friendlyError(e), Toast.LENGTH_SHORT).show());
            }
        });
    }

    private void deleteCategory(LRRCategory category) {
        new AlertDialog.Builder(getEHContext())
                .setTitle(R.string.lrr_category_action_delete)
                .setMessage(getString(R.string.lrr_category_delete_confirm, category.name))
                .setPositiveButton(android.R.string.ok, (dialog, which) -> {
                    IoThreadPoolExecutor.Companion.getInstance().execute(() -> {
                        try {
                            Context ctx = getEHContext();
                            if (ctx == null) return;
                            String serverUrl = LRRAuthManager.getServerUrl();
                            OkHttpClient client = ServiceRegistry.INSTANCE.getNetworkModule().getOkHttpClient();
                            LRRCoroutineHelper.runSuspend(
                                    (scope, cont) -> LRRCategoryApi.deleteCategory(client, serverUrl, category.id, cont));
                            if (getActivity() == null) return;
                            getActivity().runOnUiThread(() -> {
                                Toast.makeText(getEHContext(), R.string.lrr_category_deleted, Toast.LENGTH_SHORT).show();
                                fetchCategories();
                            });
                        } catch (Exception e) {
                            Log.e(TAG, "Failed to delete category", e);
                            if (getActivity() == null) return;
                            getActivity().runOnUiThread(() ->
                                    Toast.makeText(getEHContext(), LRRApiUtilsKt.friendlyError(e), Toast.LENGTH_SHORT).show());
                        }
                    });
                })
                .setNegativeButton(android.R.string.cancel, null)
                .show();
    }

    /**
     * Show long-press action menu for a category item.
     */
    private void showCategoryActions(LRRCategory category) {
        if (getEHContext() == null) return;

        String pinAction = category.isPinned()
                ? getString(R.string.lrr_category_action_unpin)
                : getString(R.string.lrr_category_action_pin);

        String[] items = {
                getString(R.string.lrr_category_action_edit),
                getString(R.string.lrr_category_action_delete),
                pinAction
        };

        new AlertDialog.Builder(getEHContext())
                .setTitle(category.name)
                .setItems(items, (dialog, which) -> {
                    switch (which) {
                        case 0: // Edit
                            showCategoryDialog(category);
                            break;
                        case 1: // Delete
                            deleteCategory(category);
                            break;
                        case 2: // Pin/Unpin
                            updateCategory(category.id, category.name,
                                    category.search, !category.isPinned());
                            break;
                    }
                })
                .show();
    }

    // ==================== View Helpers ====================

    private void showProgress() {
        if (mProgress != null) mProgress.setVisibility(View.VISIBLE);
        if (mRecyclerView != null) mRecyclerView.setVisibility(View.GONE);
        if (mErrorView != null) mErrorView.setVisibility(View.GONE);
    }

    private void showList() {
        if (mProgress != null) mProgress.setVisibility(View.GONE);
        if (mRecyclerView != null) mRecyclerView.setVisibility(View.VISIBLE);
        if (mErrorView != null) mErrorView.setVisibility(View.GONE);
    }

    private void showError(String message) {
        if (mProgress != null) mProgress.setVisibility(View.GONE);
        if (mRecyclerView != null) mRecyclerView.setVisibility(View.GONE);
        if (mErrorView != null) {
            mErrorView.setVisibility(View.VISIBLE);
            if (mErrorIcon != null) mErrorIcon.setVisibility(View.VISIBLE);
            if (mErrorTitle != null) mErrorTitle.setText(R.string.lrr_error_title);
            if (mErrorMessage != null) {
                mErrorMessage.setVisibility(View.VISIBLE);
                mErrorMessage.setText(message);
            }
            if (mErrorRetry != null) {
                mErrorRetry.setVisibility(View.VISIBLE);
                mErrorRetry.setOnClickListener(v -> fetchCategories());
            }
        }
    }

    private void showEmpty(String message) {
        if (mProgress != null) mProgress.setVisibility(View.GONE);
        if (mRecyclerView != null) mRecyclerView.setVisibility(View.GONE);
        if (mErrorView != null) {
            mErrorView.setVisibility(View.VISIBLE);
            if (mErrorIcon != null) mErrorIcon.setVisibility(View.GONE);
            if (mErrorTitle != null) mErrorTitle.setText(message);
            if (mErrorMessage != null) mErrorMessage.setVisibility(View.GONE);
            if (mErrorRetry != null) mErrorRetry.setVisibility(View.GONE);
        }
    }

    private void onCategoryClick(LRRCategory category) {
        Bundle args = new Bundle();
        args.putString(GalleryListScene.KEY_ACTION, GalleryListScene.ACTION_LIST_URL_BUILDER);

        ListUrlBuilder builder = new ListUrlBuilder();
        builder.setMode(ListUrlBuilder.MODE_NORMAL);

        if (category.isDynamic()) {
            // Dynamic category: use its search query as keyword
            builder.setKeyword(category.search.trim());
        } else {
            // Static category: use category ID filter
            builder.setKeyword("category:" + category.id);
        }

        args.putParcelable(GalleryListScene.KEY_LIST_URL_BUILDER, builder);
        startScene(new Announcer(GalleryListScene.class).setArgs(args));
    }

    // ==================== Adapter ====================

    private class CategoryAdapter extends RecyclerView.Adapter<CategoryViewHolder> {

        @NonNull
        @Override
        public CategoryViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.item_lrr_category, parent, false);
            return new CategoryViewHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull CategoryViewHolder holder, int position) {
            LRRCategory cat = mCategories.get(position);

            // Name (with pin indicator)
            if (cat.isPinned()) {
                holder.name.setText(getString(R.string.lrr_category_pinned, cat.name));
            } else {
                holder.name.setText(cat.name);
            }

            // Archive count — dynamic categories show "动态" label
            if (cat.isDynamic()) {
                holder.count.setText(getString(R.string.lrr_category_label_dynamic));
            } else {
                int archiveCount = cat.archives != null ? cat.archives.size() : 0;
                holder.count.setText(getString(R.string.lrr_category_archives, archiveCount));
            }
            holder.search.setVisibility(View.GONE);

            // Uniform icon
            holder.icon.setImageResource(R.drawable.v_heart_primary_x48);

            // Click to browse
            holder.itemView.setOnClickListener(v -> onCategoryClick(cat));

            // Long-press for actions menu
            holder.itemView.setOnLongClickListener(v -> {
                showCategoryActions(cat);
                return true;
            });
        }

        @Override
        public int getItemCount() {
            return mCategories.size();
        }
    }

    private static class CategoryViewHolder extends RecyclerView.ViewHolder {
        final ImageView icon;
        final TextView name;
        final TextView count;
        final TextView search;

        CategoryViewHolder(View itemView) {
            super(itemView);
            icon = itemView.findViewById(R.id.category_icon);
            name = itemView.findViewById(R.id.category_name);
            count = itemView.findViewById(R.id.category_count);
            search = itemView.findViewById(R.id.category_search);
        }
    }
}

File: app/src/main/java/com/hippo/ehviewer/ui/scene/LRRCategoriesScene.kt
package com.hippo.ehviewer.ui.scene

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.ImageView
import android.widget.RadioGroup
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.appbar.MaterialToolbar
import com.google.android.material.switchmaterial.SwitchMaterial
import com.hippo.ehviewer.R
import com.hippo.ehviewer.ServiceRegistry
import com.hippo.ehviewer.client.data.ListUrlBuilder
import com.hippo.ehviewer.client.lrr.LRRAuthManager
import com.hippo.ehviewer.client.lrr.LRRCategoryApi
import com.hippo.ehviewer.client.lrr.data.LRRCategory
import com.hippo.ehviewer.client.lrr.friendlyError
import com.hippo.ehviewer.client.lrr.runSuspend
import com.hippo.ehviewer.ui.scene.gallery.list.GalleryListScene
import com.hippo.scene.Announcer
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * Scene that displays LANraragi categories with full CRUD support.
 * Pinned categories appear first. Clicking a category navigates
 * to GalleryListScene filtered by that category.
 */
class LRRCategoriesScene : BaseScene() {

    private var mRecyclerView: RecyclerView? = null
    private var mProgress: View? = null
    private var mErrorView: View? = null
    private var mErrorIcon: ImageView? = null
    private var mErrorTitle: TextView? = null
    private var mErrorMessage: TextView? = null
    private var mErrorRetry: Button? = null
    private var mToolbar: MaterialToolbar? = null

    private val mCategories: MutableList<LRRCategory> = mutableListOf()
    private var mAdapter: CategoryAdapter? = null

    // Snapshot of the list last dispatched to the adapter. Read/written ONLY by
    // fetchCategories() (the single dispatch path — every CRUD op re-fetches).
    // See docs/diffutil-root-cause-analysis.md for the snapshot ownership rule.
    private var mLastSnapshot: List<LRRCategory> = emptyList()

    override fun getNavCheckedItem(): Int = R.id.nav_favourite

    override fun onCreateView2(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.scene_lrr_categories, container, false)

        mToolbar = view.findViewById(R.id.toolbar)
        mProgress = view.findViewById(R.id.progress)
        mErrorView = view.findViewById(R.id.error_view)
        mErrorIcon = view.findViewById(R.id.error_icon)
        mErrorTitle = view.findViewById(R.id.error_title)
        mErrorMessage = view.findViewById(R.id.error_message)
        mErrorRetry = view.findViewById(R.id.error_retry)
        mRecyclerView = view.findViewById(R.id.recycler_view)

        mToolbar?.apply {
            setTitle(R.string.lrr_categories_title)
            setNavigationIcon(R.drawable.v_arrow_left_dark_x24)
            setNavigationOnClickListener { onBackPressed() }
            inflateMenu(R.menu.scene_lrr_categories)
            setOnMenuItemClickListener { item ->
                if (item.itemId == R.id.action_add_category) {
                    showCategoryDialog(null)
                    true
                } else {
                    false
                }
            }
        }

        mAdapter = CategoryAdapter()
        mRecyclerView?.apply {
            layoutManager = LinearLayoutManager(context)
            setHasFixedSize(true)
            adapter = mAdapter
        }

        fetchCategories()

        return view
    }

    override fun onDestroyView() {
        super.onDestroyView()
        mRecyclerView = null
        mProgress = null
        mErrorView = null
        mErrorIcon = null
        mErrorTitle = null
        mErrorMessage = null
        mErrorRetry = null
        mToolbar = null
        mAdapter = null
        // Reset snapshot so the next view recreation starts from a clean baseline.
        mLastSnapshot = emptyList()
    }

    // ==================== Data Loading ====================

    private fun fetchCategories() {
        showProgress()

        lifecycleScope.launch(Dispatchers.IO) {
            try {
                val ctx = ehContext ?: return@launch
                val serverUrl = LRRAuthManager.getServerUrl() ?: return@launch
                val client = ServiceRegistry.networkModule.okHttpClient

                val categories = runSuspend {
                    LRRCategoryApi.getCategories(client, serverUrl)
                }

                // Sort: pinned first, then by name
                val pinned = mutableListOf<LRRCategory>()
                val unpinned = mutableListOf<LRRCategory>()
                for (cat in categories) {
                    if (cat.name.isNullOrEmpty()) continue
                    if (cat.isPinned()) {
                        pinned.add(cat)
                    } else {
                        unpinned.add(cat)
                    }
                }
                pinned.addAll(unpinned)

                activity?.runOnUiThread {
                    // Compute diff against the previously dispatched snapshot
                    // BEFORE mutating mCategories in place. Then swap in the
                    // new contents and update the snapshot, then dispatch.
                    val newList = ArrayList(pinned)
                    val adapter = mAdapter
                    if (adapter != null) {
                        val diff = DiffUtil.calculateDiff(
                            LRRCategoryDiffCallback(mLastSnapshot, newList)
                        )
                        mCategories.clear()
                        mCategories.addAll(newList)
                        mLastSnapshot = newList
                        diff.dispatchUpdatesTo(adapter)
                    } else {
                        mCategories.clear()
                        mCategories.addAll(newList)
                        mLastSnapshot = newList
                    }
                    if (mCategories.isEmpty()) {
                        showEmpty(getString(R.string.lrr_categories_empty))
                    } else {
                        showList()
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to load categories", e)
                activity?.runOnUiThread {
                    showError(friendlyError(ehContext ?: return@runOnUiThread, e))
                }
            }
        }
    }

    // ==================== CRUD Operations ====================

    /**
     * Show create/edit category dialog.
     * @param category null for create, non-null for edit.
     */
    private fun showCategoryDialog(category: LRRCategory?) {
        val ctx = ehContext ?: return
        val isEdit = category != null

        val dialogView = LayoutInflater.from(ctx).inflate(R.layout.dialog_category_edit, null)
        val nameInput = dialogView.findViewById<EditText>(R.id.category_name)
        val typeGroup = dialogView.findViewById<RadioGroup>(R.id.category_type)
        val searchLayout = dialogView.findViewById<View>(R.id.search_layout)
        val searchInput = dialogView.findViewById<EditText>(R.id.category_search)
        val pinnedSwitch = dialogView.findViewById<SwitchMaterial>(R.id.category_pinned)

        // Pre-fill for edit mode
        if (category != null) {
            nameInput.setText(category.name)
            pinnedSwitch.isChecked = category.isPinned()
            if (category.isDynamic()) {
                typeGroup.check(R.id.type_dynamic)
                searchLayout.visibility = View.VISIBLE
                searchInput.setText(category.search)
            }
            // Disable type switch for edit (can't change static<->dynamic if archives exist)
            if (category.archives.isNotEmpty()) {
                dialogView.findViewById<View>(R.id.type_static).isEnabled = false
                dialogView.findViewById<View>(R.id.type_dynamic).isEnabled = false
            }
        }

        // Toggle search field visibility
        typeGroup.setOnCheckedChangeListener { _, id ->
            searchLayout.visibility = if (id == R.id.type_dynamic) View.VISIBLE else View.GONE
        }

        AlertDialog.Builder(ctx)
            .setTitle(if (isEdit) R.string.lrr_category_edit else R.string.lrr_category_create)
            .setView(dialogView)
            .setPositiveButton(android.R.string.ok) { _, _ ->
                val name = nameInput.text.toString().trim()
                if (name.isEmpty()) {
                    Toast.makeText(ctx, R.string.lrr_category_name_empty, Toast.LENGTH_SHORT).show()
                    return@setPositiveButton
                }
                val isDynamic = typeGroup.checkedRadioButtonId == R.id.type_dynamic
                val search = if (isDynamic) searchInput.text.toString().trim() else null
                val pinned = pinnedSwitch.isChecked

                if (category != null) {
                    updateCategory(category.id!!, name, search, pinned)
                } else {
                    createCategory(name, search, pinned)
                }
            }
            .setNegativeButton(android.R.string.cancel, null)
            .show()
    }

    private fun createCategory(name: String, search: String?, pinned: Boolean) {
        lifecycleScope.launch(Dispatchers.IO) {
            try {
                val ctx = ehContext ?: return@launch
                val serverUrl = LRRAuthManager.getServerUrl() ?: return@launch
                val client = ServiceRegistry.networkModule.okHttpClient
                runSuspend {
                    LRRCategoryApi.createCategory(client, serverUrl, name, search, pinned)
                }
                activity?.runOnUiThread {
                    Toast.makeText(ehContext, R.string.lrr_category_created, Toast.LENGTH_SHORT).show()
                    fetchCategories()
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to create category", e)
                activity?.runOnUiThread {
                    Toast.makeText(
                        ehContext,
                        friendlyError(ehContext ?: return@runOnUiThread, e),
                        Toast.LENGTH_SHORT
                    ).show()
                }
            }
        }
    }

    private fun updateCategory(categoryId: String, name: String, search: String?, pinned: Boolean) {
        lifecycleScope.launch(Dispatchers.IO) {
            try {
                val ctx = ehContext ?: return@launch
                val serverUrl = LRRAuthManager.getServerUrl() ?: return@launch
                val client = ServiceRegistry.networkModule.okHttpClient
                runSuspend {
                    LRRCategoryApi.updateCategory(client, serverUrl, categoryId, name, search, pinned)
                }
                activity?.runOnUiThread {
                    Toast.makeText(ehContext, R.string.lrr_category_updated, Toast.LENGTH_SHORT).show()
                    fetchCategories()
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to update category", e)
                activity?.runOnUiThread {
                    Toast.makeText(
                        ehContext,
                        friendlyError(ehContext ?: return@runOnUiThread, e),
                        Toast.LENGTH_SHORT
                    ).show()
                }
            }
        }
    }

    private fun deleteCategory(category: LRRCategory) {
        val ctx = ehContext ?: return
        AlertDialog.Builder(ctx)
            .setTitle(R.string.lrr_category_action_delete)
            .setMessage(getString(R.string.lrr_category_delete_confirm, category.name))
            .setPositiveButton(android.R.string.ok) { _, _ ->
                lifecycleScope.launch(Dispatchers.IO) {
                    try {
                        val innerCtx = ehContext ?: return@launch
                        val serverUrl = LRRAuthManager.getServerUrl() ?: return@launch
                        val client = ServiceRegistry.networkModule.okHttpClient
                        runSuspend {
                            LRRCategoryApi.deleteCategory(client, serverUrl, category.id!!)
                        }
                        activity?.runOnUiThread {
                            Toast.makeText(ehContext, R.string.lrr_category_deleted, Toast.LENGTH_SHORT).show()
                            fetchCategories()
                        }
                    } catch (e: Exception) {
                        Log.e(TAG, "Failed to delete category", e)
                        activity?.runOnUiThread {
                            Toast.makeText(
                                ehContext,
                                friendlyError(ehContext ?: return@runOnUiThread, e),
                                Toast.LENGTH_SHORT
                            ).show()
                        }
                    }
                }
            }
            .setNegativeButton(android.R.string.cancel, null)
            .show()
    }

    /**
     * Show long-press action menu for a category item.
     */
    private fun showCategoryActions(category: LRRCategory) {
        val ctx = ehContext ?: return

        val pinAction = if (category.isPinned()) {
            getString(R.string.lrr_category_action_unpin)
        } else {
            getString(R.string.lrr_category_action_pin)
        }

        val items = arrayOf(
            getString(R.string.lrr_category_action_edit),
            getString(R.string.lrr_category_action_delete),
            pinAction
        )

        AlertDialog.Builder(ctx)
            .setTitle(category.name)
            .setItems(items) { _, which ->
                when (which) {
                    0 -> showCategoryDialog(category) // Edit
                    1 -> deleteCategory(category) // Delete
                    2 -> updateCategory( // Pin/Unpin
                        category.id!!,
                        category.name!!,
                        category.search,
                        !category.isPinned()
                    )
                }
            }
            .show()
    }

    // ==================== View Helpers ====================

    private fun showProgress() {
        mProgress?.visibility = View.VISIBLE
        mRecyclerView?.visibility = View.GONE
        mErrorView?.visibility = View.GONE
    }

    private fun showList() {
        mProgress?.visibility = View.GONE
        mRecyclerView?.visibility = View.VISIBLE
        mErrorView?.visibility = View.GONE
    }

    private fun showError(message: String) {
        mProgress?.visibility = View.GONE
        mRecyclerView?.visibility = View.GONE
        mErrorView?.apply {
            visibility = View.VISIBLE
            mErrorIcon?.visibility = View.VISIBLE
            mErrorTitle?.setText(R.string.lrr_error_title)
            mErrorMessage?.apply {
                visibility = View.VISIBLE
                text = message
            }
            mErrorRetry?.apply {
                visibility = View.VISIBLE
                setOnClickListener { fetchCategories() }
            }
        }
    }

    private fun showEmpty(message: String) {
        mProgress?.visibility = View.GONE
        mRecyclerView?.visibility = View.GONE
        mErrorView?.apply {
            visibility = View.VISIBLE
            mErrorIcon?.visibility = View.GONE
            mErrorTitle?.text = message
            mErrorMessage?.visibility = View.GONE
            mErrorRetry?.visibility = View.GONE
        }
    }

    private fun onCategoryClick(category: LRRCategory) {
        val args = Bundle().apply {
            putString(GalleryListScene.KEY_ACTION, GalleryListScene.ACTION_LIST_URL_BUILDER)
        }

        val builder = ListUrlBuilder().apply {
            mode = ListUrlBuilder.MODE_NORMAL
            keyword = if (category.isDynamic()) {
                // Dynamic category: use its search query as keyword
                category.search!!.trim()
            } else {
                // Static category: use category ID filter
                "category:${category.id}"
            }
        }

        args.putParcelable(GalleryListScene.KEY_LIST_URL_BUILDER, builder)
        startScene(Announcer(GalleryListScene::class.java).setArgs(args))
    }

    // ==================== Adapter ====================

    private inner class CategoryAdapter : RecyclerView.Adapter<CategoryViewHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CategoryViewHolder {
            val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.item_lrr_category, parent, false)
            return CategoryViewHolder(view)
        }

        override fun onBindViewHolder(holder: CategoryViewHolder, position: Int) {
            val cat = mCategories[position]

            // Name (with pin indicator)
            if (cat.isPinned()) {
                holder.name.text = getString(R.string.lrr_category_pinned, cat.name)
            } else {
                holder.name.text = cat.name
            }

            // Archive count -- dynamic categories show "动态" label
            if (cat.isDynamic()) {
                holder.count.text = getString(R.string.lrr_category_label_dynamic)
            } else {
                val archiveCount = cat.archives.size
                holder.count.text = getString(R.string.lrr_category_archives, archiveCount)
            }
            holder.search.visibility = View.GONE

            // Uniform icon
            holder.icon.setImageResource(R.drawable.v_heart_primary_x48)

            // Click to browse
            holder.itemView.setOnClickListener { onCategoryClick(cat) }

            // Long-press for actions menu
            holder.itemView.setOnLongClickListener {
                showCategoryActions(cat)
                true
            }
        }

        override fun getItemCount(): Int = mCategories.size
    }

    private class CategoryViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val icon: ImageView = itemView.findViewById(R.id.category_icon)
        val name: TextView = itemView.findViewById(R.id.category_name)
        val count: TextView = itemView.findViewById(R.id.category_count)
        val search: TextView = itemView.findViewById(R.id.category_search)
    }

    /**
     * DiffUtil callback for LRRCategory lists. Identity is the category `id`
     * (LANraragi's CAT-prefixed string id). Content compares everything that
     * onBindViewHolder reads so the row repaints when name/pinned/search/
     * archive count change.
     */
    private class LRRCategoryDiffCallback(
        private val oldList: List<LRRCategory>,
        private val newList: List<LRRCategory>
    ) : DiffUtil.Callback() {

        override fun getOldListSize(): Int = oldList.size
        override fun getNewListSize(): Int = newList.size

        override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            return oldList[oldItemPosition].id == newList[newItemPosition].id
        }

        override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            val o = oldList[oldItemPosition]
            val n = newList[newItemPosition]
            return o.name == n.name &&
                o.isPinned() == n.isPinned() &&
                o.isDynamic() == n.isDynamic() &&
                o.search == n.search &&
                o.archives.size == n.archives.size
        }
    }

    companion object {
        private const val TAG = "LRRCategoriesScene"
    }
}
